src/app_model/backends/qt/_qaction.py

```diff
@@ -118,52 +120,70 @@ class QMenuItemAction(QCommandRuleAction):
-        app = Application.get_or_create(app) if isinstance(app, str) else app
-        key = (id(app), hash(menu_item))
-        if cache and key in cls._cache:
-            return cls._cache[key]
+    Parameters
+    ----------
+    menu_item : MenuItem
+        `MenuItem` instance to create an action for.
+    app : Application | str
+        Application instance or name of application instance.
+    parent : QObject | None
+        Optional parent widget, by default None
+    """
 
-        self = super().__new__(cls)
-        if cache:
-            cls._cache[key] = self
-        return self  # type: ignore [no-any-return]
+    _cache: ClassVar[WeakValueDictionary[tuple[int, int], QMenuItemAction]] = (
+        WeakValueDictionary()
+    )
 
     def __init__(
         self,
         menu_item: MenuItem,
         app: Application | str,
         parent: QObject | None = None,
-        *,
-        cache: bool = True,  # used in __new__
     ):
-        initialized = False
-        with contextlib.suppress(RuntimeError):
-            initialized = getattr(self, "_initialized", False)
-
-        if not initialized:
-            super().__init__(menu_item.command, app, parent)
-            self._menu_item = menu_item
-            key = (id(self._app), hash(menu_item))
-            self.destroyed.connect(lambda: QMenuItemAction._cache.pop(key, None))
-            self._app.destroyed.connect(lambda: QMenuItemAction._cache.pop(key, None))
-            self._initialized = True
-
+        super().__init__(menu_item.command, app, parent)
+        self._menu_item = menu_item
         with contextlib.suppress(NameError):
             self.update_from_context(self._app.context)
 
+    @staticmethod
+    def _cache_key(app: Application, menu_item: MenuItem) -> tuple[int, int]:
+        return (id(app), hash(menu_item))
+
+    def _remove_from_cache(self) -> None:
+        cache_key = QMenuItemAction._cache_key(self._app, self._menu_item)
+        QMenuItemAction._cache.pop(cache_key, None)
+
+    @classmethod
+    def create(
+        cls,
+        menu_item: MenuItem,
+        app: Application | str,
+        parent: QObject | None = None,
+    ) -> Self:
+        """Create a new QMenuItemAction for the given menu item.
+
+        Prefer this method over `__init__` to ensure that the cache is used,
+        so that:
+
+        ```python
+        a1 = QMenuItemAction.create(action, full_app)
+        a2 = QMenuItemAction.create(action, full_app)
+        a1 is a2  # True
+        ```
+        """
+        app = Application.get_or_create(app) if isinstance(app, str) else app
+        cache_key = QMenuItemAction._cache_key(app, menu_item)
+        if cache_key in cls._cache:
+            res = cls._cache[cache_key]
+            res.setParent(parent)
+            return res
+
+        cls._cache[cache_key] = obj = cls(menu_item, app, parent)
+        obj.destroyed.connect(obj._remove_from_cache)
+        app.destroyed.connect(obj._remove_from_cache)
```

Comment:
```suggestion
```

Reply:
sure, if you want to!  I agree, they seem kinda useless at this point and perhaps the weak dict is sufficient anyway